core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AddDynamicFilterSource.java

```diff
@@ -0,0 +1,198 @@
+
+        @Override
+        public Result apply(JoinNode joinNode, Captures captures, Context context)
+        {
+            Optional<PlanNode> rewritten = recurseAndRewriteExchangeSource(
+                    captures.get(BUILD_SIDE_NODE),
+                    context.getIdAllocator(),
+                    context.getLookup(),
+                    joinNode.getDynamicFilters());
+            if (rewritten.isEmpty()) {
+                return Result.ofPlanNode(joinNode.withoutDynamicFilters());
+            }
+
+            return Result.ofPlanNode(joinNode.withoutDynamicFilters()
+                    .replaceChildren(ImmutableList.of(joinNode.getLeft(), rewritten.get())));
+        }
+    }
+
+    public static class AddSemiJoinDynamicFilterSource
+            implements Rule<SemiJoinNode>
+    {
+        private static final Capture<PlanNode> FILTERING_SOURCE_NODE = Capture.newCapture();
+        private static final Pattern<SemiJoinNode> PATTERN = semiJoin()
+                .matching(node -> node.getDynamicFilterId().isPresent())
+                .with(getFilteringSource().capturedAs(FILTERING_SOURCE_NODE));
+
+        @Override
+        public Pattern<SemiJoinNode> getPattern()
+        {
+            return PATTERN;
+        }
+
+        @Override
+        public boolean isEnabled(Session session)
+        {
+            return getRetryPolicy(session) == RetryPolicy.TASK;
+        }
+
+        @Override
+        public Result apply(SemiJoinNode semiJoinNode, Captures captures, Context context)
+        {
+            Optional<PlanNode> rewritten = recurseAndRewriteExchangeSource(
+                    captures.get(FILTERING_SOURCE_NODE),
+                    context.getIdAllocator(),
+                    context.getLookup(),
+                    ImmutableMap.of(semiJoinNode.getDynamicFilterId().orElseThrow(), semiJoinNode.getFilteringSourceJoinSymbol()));
+            if (rewritten.isEmpty()) {
+                return Result.ofPlanNode(semiJoinNode.withoutDynamicFilter());
+            }
+
+            return Result.ofPlanNode(semiJoinNode.withoutDynamicFilter()
+                    .replaceChildren(ImmutableList.of(semiJoinNode.getSource(), rewritten.get())));
+        }
+    }
+
+    private static Optional<PlanNode> recurseAndRewriteExchangeSource(
+            PlanNode node,
+            PlanNodeIdAllocator idAllocator,
+            Lookup lookup,
+            Map<DynamicFilterId, Symbol> dynamicFilters)
+    {
+        // TODO: add support for cases where the build side has multiple sources, e.g., a UNION ALL for a partitioned join
+        if (node.getSources().size() != 1) {
+            return Optional.empty();
+        }
+        if (isRemoteExchange(node)) {
+            ExchangeNode exchangeNode = (ExchangeNode) node;
+            return Optional.of(rewriteRemoteExchangeSource(exchangeNode, idAllocator, dynamicFilters));
+        }
+
+        if (node instanceof ProjectNode || isLocalRepartitionOrGatherExchange(node)) {
```

Comment:
In what cases there is a `ProjectNode`?

Reply:
I didn't find an actual query where there is a ProjectNode here. But I also wasn't sure whether this can ever happen.
I'm also thinking that for rewrite of dynamic filter source to be correct, the projection here should be an identity projection. Does it make sense to keep this with a check that it's identity projection or should we remove it entirely ?

Reply:
I found the relevant query now
```
explain SELECT l.orderkey FROM lineitem l, orders o where  l.orderkey >= o.orderkey and l.orderkey <= o.custkey + 1;

Fragment 1 [tpch:orders:15000]
     Output layout: [orderkey]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     FilterProject[filterPredicate = (("orderkey" >= "orderkey_1") AND ("orderkey" <= "expr"))]
     │   Layout: [orderkey:bigint]
     │   Estimates: {rows: 5641500 (48.42MB), cpu: 45.40G, memory: 263.67kB, network: 263.67kB}/{rows: 5641500 (48.42MB), cpu: 45.44G, memory: 263.67kB, network: 263.67kB}
     └─ CrossJoin
        │   Layout: [orderkey:bigint, orderkey_1:bigint, expr:bigint]
        │   Estimates: {rows: 902625000 (22.70GB), cpu: 22.70G, memory: 263.67kB, network: 263.67kB}
        │   Distribution: REPLICATED
        ├─ TableScan[tpch:tiny:lineitem, grouped = false]
        │      Layout: [orderkey:bigint]
        │      Estimates: {rows: 60175 (528.88kB), cpu: 528.88k, memory: 0B, network: 0B}
        │      orderkey := tpch:orderkey
        └─ LocalExchange[SINGLE] ()
           │   Layout: [expr:bigint, orderkey_1:bigint]
           │   Estimates: {rows: 15000 (263.67kB), cpu: 527.34k, memory: 0B, network: 263.67kB}
           └─ Project[]
              │   Layout: [expr:bigint, orderkey_1:bigint]
              │   Estimates: {rows: 15000 (263.67kB), cpu: 527.34k, memory: 0B, network: 263.67kB}
              │   expr := ("custkey" + BIGINT '1')
              └─ RemoteSource[2]
                     Layout: [orderkey_1:bigint, custkey:bigint]

 Fragment 2 [tpch:orders:15000]
     Output layout: [orderkey_1, custkey]
     Output partitioning: BROADCAST []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     TableScan[tpch:tiny:orders, grouped = false]
         Layout: [orderkey_1:bigint, custkey:bigint]
         Estimates: {rows: 15000 (263.67kB), cpu: 263.67k, memory: 0B, network: 0B}
         custkey := tpch:custkey
         orderkey_1 := tpch:orderkey
```
In this case we can't move the DynamicFilterSource beyond the ProjectNode. In this PR I'll make sure that we check for identity projection and remove the DF in task retry mode, we can tackle supporting it in a follow-up.
In general, should this project have been moved into Fragment 2 by some rule ?